crates/p2p/src/swarm/mod.rs

```diff
@@ -387,13 +492,37 @@ impl P2P {
                 debug!("Publishing message");
                 trace!(?msg, "Publishing message");
 
-                // TODO(Velnbur): add retry mechanism later, instead of skipping the error
-                let _ = self
+                let message_id = self
                     .swarm
                     .behaviour_mut()
                     .gossipsub
                     .publish(TOPIC.hash(), msg.into_raw().encode_to_vec())
-                    .inspect_err(|err| error!(%err, "Failed to publish msg through gossipsub"));
+                    .inspect_err(|err| {
+                        match err {
+                            PublishError::Duplicate => {
+                                warn!(%err, "Failed to publish msg through gossipsub, message already exists");
+                            }
+                            PublishError::SigningError(signing_error) => {
+                                error!(%signing_error, "Failed to sign message");
+                            }
+                            PublishError::InsufficientPeers => {
+                                error!("Insufficient peers to publish message");
+                            }
+                            PublishError::MessageTooLarge => {
+                                error!("Message too large to publish");
+                            }
+                            PublishError::TransformFailed(error) => {
+                                error!(%error, "Failed to transform message");
+                            }
+                            PublishError::AllQueuesFull(num_peers_attempted) => {
+                                error!(%num_peers_attempted, "All queues full, dropping message");
+                            }
+                        }
+                    });
+
+                if message_id.is_ok() {
+                    debug!(message_id=%message_id.unwrap(), "Message published");
+                }
```

Comment:
Just return here.

```suggestion
                    })?;
                    
                    // unwrap() is ok here because of the early return above
                    debug!(message_id=%message_id.unwrap(), "Message published");
```

(indentation might be wrong on this suggestion).